src/persist-txn/src/txn_cache.rs

```diff
@@ -996,127 +996,118 @@ mod tests {
+        // of events:
+        //
+        // - Registrations at: [2,8], [12,13]
+        // - Direct writes at: 1, 10
+        // - Writes via txns at: 4, 5, 7
+        let mut c = TxnsCacheState::new(ShardId::new(), 0, None);
+
+        // empty
+        assert_eq!(c.progress_exclusive, 0);
+        assert!(std::panic::catch_unwind(|| c.data_snapshot(d0, 0)).is_err());
+        assert_eq!(c.data_listen_next(&d0, 0), WaitForTxnsProgress);
+
+        // ts 0 (never registered)
+        testcase(&mut c, 0, d0, ds(None, 0, 1), ReadDataTo(1));
+
+        // ts 1 (direct write)
+        testcase(&mut c, 1, d0, ds(None, 1, 2), ReadDataTo(2));
+
+        // ts 2 (register)
+        c.push_register(d0, 2, 1);
+        testcase(&mut c, 2, d0, ds(None, 2, 3), EmitLogicalProgress(3));
+
+        // ts 3 (registered, not written)
+        testcase(&mut c, 3, d0, ds(None, 3, 4), EmitLogicalProgress(4));
+
+        // ts 4 (written via txns)
+        c.push_append(d0, vec![4], 4, 1);
+        testcase(&mut c, 4, d0, ds(Some(4), 4, 5), ReadDataTo(5));
+
+        // ts 5 (written via txns, write at preceding ts)
+        c.push_append(d0, vec![5], 5, 1);
+        testcase(&mut c, 5, d0, ds(Some(5), 5, 6), ReadDataTo(6));
+
+        // ts 6 (registered, not written, write at preceding ts)
+        testcase(&mut c, 6, d0, ds(Some(5), 6, 7), EmitLogicalProgress(7));
+
+        // ts 7 (written via txns, write at non-preceding ts)
+        c.push_append(d0, vec![7], 7, 1);
+        testcase(&mut c, 7, d0, ds(Some(7), 7, 8), ReadDataTo(8));
+
+        // ts 8 (forget). Forget guarantees that all writes are applied, so no
+        // latest_write. It also currently is inclusive, so we can emit logical
+        // progress, though we might want to make the interval half-open?
+        c.push_register(d0, 2, -1);
+        testcase(&mut c, 8, d0, ds(None, 8, 9), EmitLogicalProgress(9));
+
+        // ts 9 (not registered, not written). This ReadDataTo would block until
+        // the write happens at ts 10.
+        testcase(&mut c, 9, d0, ds(None, 9, 10), ReadDataTo(10));
+
+        // ts 10 (written directly)
+        testcase(&mut c, 10, d0, ds(None, 10, 11), ReadDataTo(11));
+
+        // ts 11 (not registered, not written) This ReadDataTo would block until
+        // the register happens at 12.
+        testcase(&mut c, 11, d0, ds(None, 11, 12), ReadDataTo(12));
+
+        // ts 12 (registered, previously forgotten)
+        c.push_register(d0, 12, 1);
+        testcase(&mut c, 12, d0, ds(None, 12, 13), EmitLogicalProgress(13));
+
+        // ts 13 (forgotten, registered at preceding ts)
+        c.push_register(d0, 12, -1);
+        testcase(&mut c, 13, d0, ds(None, 13, 14), ReadDataTo(14));
+
+        // Now that we have more history, some of the old answers change! In
+        // particular, we can ReadDataTo much later times.
+        assert_eq!(c.data_listen_next(&d0, 0), ReadDataTo(12));
```

Comment:
The current impl returns `ReadDataTo(8)`, which is also correct and definitely pessimistic.

Reply:
This is also related to https://github.com/MaterializeInc/materialize/pull/23911/files#r1566264250